### PR TITLE
Enable journalists to upload their GPG public key

### PR DESCRIPTION
**Pull request by redshiftzero**
This is a first step toward #98 (enabling sources to submit to specific journalists). In order to have sources select the journalists that they want to send their document to, we need in the SD database 1) more than just a username to show sources: probably we need for each journalist at least their name and a brief blurb about what they work on and 2) we need their public keys. 

This PR adds these to the database and adds a couple of new routes enabling a journalist to submit: their PGP key, full name, and a bit of profile text that they can use to describe themselves to potential sources (e.g. "Crime Reporter", "National Security Reporter", etc.). The intention here is that this information (name and blurb) would eventually be displayed to sources that want to send documents directly to a specific person and then on the backend documents can be encrypted to the journalist's public key. Journalists are able to get to their profiles by clicking the new link "Profile" in the upper right corner of the interface:

profile-linky

Their PGP key, name, blurb now appears on the journalist interface under a new profile route `/profile/<username>`:

profile

They can click "Edit my profile" to modify any of this information:

edit-profile

Logged-in users are able to look at one another's profiles if they are so inclined (but are unable to edit them as the "Edit my profile" link will not appear). 

The profile page also displays the sources that the journalist is assigned to. This enables journalists to remind themselves what they are assigned to as well as potentially for collaboration (they can quickly see what one of their colleagues is assigned to). 

Will add more tests before merge if this general approach is useful/reasonable, thoughts/comments/questions/concerns welcome

### CODE REVIEW

**Comment by redshiftzero** [comment]
It sounded like from AFK discussion that this is a direction that is useful so I've added unit tests for it and added a feature enabling administrators to edit this information in the database in the not-that-unlikely-situation that journalists are slow in putting their profile information and GPG keys on the SecureDrop server. An administrator can just go in and add this information for the users herself by going to the admin interface and clicking on the "Edit Profile" link of the user she wishes to edit:

new-edit-profile-link

In addition, a link now appears on the profile page of any user if the person viewing the page is an administrator. The link directs them to the edit profile link for that user:

admin-link-edit-profile

(Note: I followed the same approach taken during the creation of `admin_edit_user` and `edit_user` to make the profile edit routes for regular users and administrators separate.)

**Comment by fowlslegs** [comment]
Do we want this in the 0.4 milestone?

**Comment by redshiftzero** [comment]
I'm fine with pushing this until a future release in order to keep 0.4 constrained to shipping develop.

**Comment by redshiftzero** [comment]
Hey @fowlslegs saw you were taking a stab at reviewing this. Note there is one remaining to do here (sorry, just took another look at this after having eyes off it for a couple of months): I need to add validation on `username` in `profile()`. Let me know if you notice anything else. Merge of this is a few releases out due to requiring database migrations support, but any comments either in general or detail are very welcome!

**Comment by conorsch** [comment]
> Merge of this is a few releases out due to requiring database migrations support

We should set a milestone to queue up all pending changes blocked by migration support—tentatively 0.5, could land earlier though, especially if we manage to release more rapidly, and implement a migration strategy that can be applied via automatic updates.

File: securedrop/journalist_templates/profile.html [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    </div>
+  </p>
+  {% endfor %}
+{% else %}
+<p>User is not currently assigned to any sources.</p>
+{% endif %}
+
+<p><b>PGP public key:</b></p>
+{% if user.pgp_key: %}
+<pre>{{ user.pgp_key }}</pre>
```
Comment (fowlslegs) [inline securedrop/journalist_templates/profile.html]: Maybe we don't want their entire PGP public key to be displayed here. I think a fingerprint would look nicer. We can have a "Download public key" link next to it, but journos shouldn't need each other's public keys on their JWs, so what's the point anyway?

File: securedrop/journalist.py [hunk 2]
```diff
@@ -409,6 +411,55 @@ def remove_star(sid):
...
+        user.full_name = form.full_name.data
+        user.about = form.about.data
+        user.pgp_key = form.pgp_key.data
+        db_session.commit()
+        flash('Successfully edited your profile!', 'notification')
+        return redirect(url_for('profile', username=g.user.username))
+    return render_template("edit_profile.html", form=form, user=g.user)
+
+
+@app.route('/admin/edit-profile/<int:user_id>', methods=('GET', 'POST'))
```
Comment (fowlslegs) [inline securedrop/journalist.py]: I think the `/admin/edit-profile/<username>/` URL routing scheme is nicer, but not super important.

File: securedrop/journalist.py [hunk 3]
```diff
@@ -409,6 +411,55 @@ def remove_star(sid):
...
+        form.full_name.data = user.full_name
+        form.about.data = user.about
+        form.pgp_key.data = user.pgp_key
+    elif request.method == 'POST' and form.validate():
+        user.full_name = form.full_name.data
+        user.about = form.about.data
+        user.pgp_key = form.pgp_key.data
+        db_session.commit()
+        flash('Successfully edited the profile for {}!'.format(user.username),
+              'notification')
```
Comment (fowlslegs) [inline securedrop/journalist.py]: Probably want `success` instead of `notification` here.

File: securedrop/journalist.py [hunk 4]
```diff
@@ -409,6 +411,55 @@ def remove_star(sid):
...
+        form.full_name.data = g.user.full_name
+        form.about.data = g.user.about
+        form.pgp_key.data = g.user.pgp_key
+    elif request.method == 'POST' and form.validate():
+        user = Journalist.query.filter_by(username=g.user.username).one()
+        user.full_name = form.full_name.data
+        user.about = form.about.data
+        user.pgp_key = form.pgp_key.data
+        db_session.commit()
+        flash('Successfully edited your profile!', 'notification')
```
Comment (fowlslegs) [inline securedrop/journalist.py]: Probably want `success` instead of `notification` here.

**Comment by redshiftzero** [comment]
Closing until the project is in a position to make changes like this

**Pull request closed without merging**